[ryamgoh] iP

ryamgoh opened the pull request:
# Duke

## Description

This iP PR aims to enhance our project by implementing several key features and addressing important issues. The following lists outline the changes and tasks to be completed:

### Features Added:

- [x] 🌟 Added ability to add and delete tasks
- [x] 📊 Added validation
- [x] 🛡️ Added list functionality
- [x] 🔄 Integrated persistent storage

### Issues Resolved:

1. 🐛 Fixed a critical bug causing data to fail to persist
2. 🐞 Addressed performance issues related to database queries.

### Code Changes:

```java
// Example Java code snippet with syntax highlighting
package duke;

import duke.commands.Command;
import duke.exceptions.DukeException;
import duke.parser.Parser;
import duke.storage.Storage;
import duke.tasks.TaskList;
import duke.ui.Main;
import duke.ui.Ui;
import javafx.application.Application;

public class Duke {
    /**
     * The task list
     */
    private final TaskList taskList;
    /**
     * The ui object
     */
    private final Ui ui;
    /**
     * The storage object
     */
    private final Storage storage;

    /**
     * Constructor for Duke
     *
     * @param filePath the file path
     */
    public Duke(String filePath) {
        this.storage = new Storage(filePath);
        this.taskList = new TaskList(this.storage.readData());
        this.ui = new Ui();
    }

    /**
     * Constructor for Duke
     */
    public Duke() {
        this.storage = new Storage("data/duke.txt");
        this.taskList = new TaskList(this.storage.readData());
        this.ui = new Ui();
    }

    /**
     * Initialises the program
     *
     * @param args Input args.
     */
    public static void main(String[] args) {
        Application.launch(Main.class, args);
    }

    /**
     * You should have your own function to generate a response to user input.
     * Replace this stub with your completed method.
     */
    public String getResponse(String input) {
        try {
            Command command = Parser.parse(input);
            return command.execute(this.taskList, this.ui, this.storage);
        } catch (DukeException exc) {
            return exc.getMessage();
        }
    }
}
```

### Task List:
- **Conduct a security audit to identify potential vulnerabilities.**
- **Update documentation with the latest changes.**
- **Prepare a release candidate for deployment.**

### Additional Notes:

> 
For more details, please refer to the [full PR here](https://github.com/ryamgoh/ip).

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

ryamgoh committed:
Level-0

ryamgoh committed:
Level-1

ryamgoh committed:
Level-2

ryamgoh committed:
Level-3

ryamgoh committed:
Level-4

ryamgoh committed:
A-TextUiTesting

ryamgoh committed:
Level-5

ryamgoh committed:
Level-5

ryamgoh committed:
Level-6

ryamgoh committed:
Add Storage

ryamgoh committed:
Level-7

ryamgoh committed:
Merge branch 'branch-Level-7'

ryamgoh committed:
Level-8

ryamgoh committed:
Merge branch 'branch-Level-8'

ryamgoh committed:
Refactor Code

ryamgoh committed:
A-MoreOOP

ryamgoh committed:
Merge tag 'A-MoreOOP'

ryamgoh committed:
A-Packages

ryamgoh committed:
Merge branch 'branch-Packages'

ryamgoh committed:
Merge branch 'add-gradle-support' of https://github.com/ryamgoh/ip

ryamgoh committed:
A-Gradle

ryamgoh committed:
A-JUnit

ryamgoh committed:
A-Jar

ryamgoh committed:
Javadoc

ryamgoh committed:
Refactor with Command

ryamgoh committed:
Merge branch 'branch-experimentWithRefactoring'

ryamgoh committed:
Add Javadoc

ryamgoh committed:
Merge branch 'master' into A-JavaDoc

ryamgoh committed:
A-JavaDoc

ryamgoh committed:
Merge tag 'A-JavaDoc'

ryamgoh committed:
Level-9

ryamgoh committed:
Merge branch 'branch-Level-9'

ryamgoh committed:
A-CheckStyle

ryamgoh committed:
Merge branch 'branch-A-CheckStyle'

ryamgoh committed:
Level-10

ryamgoh committed:
Merge branch 'branch-Level-10'

ryamgoh committed:
Remove redundancies

teozern1 commented on src/main/java/duke/commands/DeleteCommand.java:
```diff
@@ -0,0 +1,54 @@
+import duke.tasks.TaskList;
+import duke.ui.Ui;
+
+public class DeleteCommand extends Command {
```
Good that all the commands have the same naming convention

teozern1 commented on src/main/java/duke/storage/Storage.java:
```diff
@@ -0,0 +1,158 @@
+	/**
+	 * Creates the directory
+	 */
+	public void createDirectory() {
```
Good naming of methods using verbs

teozern1 commented on src/main/java/duke/storage/Storage.java:
```diff
@@ -0,0 +1,158 @@
+	 * @param line the string to be formatted
+	 * @return the task
+	 */
+	public Task formatStringToTask(String line) {
```
Method name is concise but descriptive enough 👍

teozern1 commented on src/main/java/duke/tasks/Deadline.java:
```diff
@@ -0,0 +1,54 @@
+	@Override
+	public String toString() {
+		return "[D]" + super.toString()
+						+ " (by: " + this.deadline.format(DateTimeFormatter.ofPattern("MMM d yyyy hh:mm a")) + ")";
```
Not sure whether this is the convention for indentation?

teozern1 commented on src/main/java/duke/tasks/TaskList.java:
```diff
@@ -0,0 +1,127 @@
+     *
+     * @return the size of the list
+     */
+    public int length() {
```
Should this be getLength?

teozern1 reviewed (commented):
I like the use of enums for parsing the commands. The code is clean with no major issues that I can spot!

lambraydon commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -0,0 +1,269 @@
+package duke.parser;
+
+import duke.commands.*;
```
Might it be better to explicitly list each import to adhere to the coding convention?

lambraydon commented on src/main/java/duke/storage/Storage.java:
```diff
@@ -0,0 +1,158 @@
+
+		DateTimeFormatter formatter = DateTimeFormatter.ofPattern("MMM d yyyy hh:mm a");
+
+		switch (type) {
+			case "T":
+				task = new Todo(description, status);
+				break;
+			case "D":
+				task = new Deadline(description, LocalDateTime.parse(split[3], formatter), status);
+				break;
+			case "E":
+				String[] interval = split[3].split(" - ", 2);
+				if (interval.length < 2) {
+					task = null;
+				} else {
+					task = new Event(description, LocalDateTime.parse(interval[0], formatter), LocalDateTime.parse(interval[1], formatter), status);
+				}
+				break;
+			default:
+				task = null;
+		}
+
+		return task;
+	}
```
Not sure whether this adheres to the coding convention of switch statements?

lambraydon commented on src/main/java/duke/tasks/Task.java:
```diff
@@ -0,0 +1,97 @@
+     *
+     * @return String description of the task.
+     */
+    public String getStatusIcon() {
```
Good naming of methods in verb form!

lambraydon reviewed (commented):
Overall LGTM! Just a few nitpicks in codestyle

ryamgoh committed:
Fix coding violations

ryamgoh committed:
Add 2 new Asserts in Task

ryamgoh committed:
Add more Assertions to Add, Storage, Commands

ryamgoh committed:
Merge pull request #1 from ryamgoh/branch-A-Assertions

Add more Assertions to Add, Storage, Commands

ryamgoh committed:
Fix coding violations in FindCommand Storage

ryamgoh committed:
Fix Code quality of Parser Class

ryamgoh committed:
Merge pull request #2 from ryamgoh/branch-A-CodeQuality

Fix Code quality of Parser Class

ryamgoh committed:
Merge pull request #2 from ryamgoh/branch-A-CodeQuality

Fix Code quality of Parser Class

Abstracted parts of Parser to improve coding quality

ryamgoh committed:
Merge branch 'master' of https://github.com/ryamgoh/ip

ryamgoh committed:
Add body to Commit Message

As per the last commit a491046, I abstracted parts of Parser class to improve code quality

ryamgoh committed:
Add Help Command as part of the BCD Extension

This is another full commit message \n Added the help command to help the user navigate the bot

ryamgoh committed:
Add some javadocs in package.ui

ryamgoh committed:
Add name of bot above

ryamgoh committed:
Change duke.txt

Change the duke.txt

ryamgoh committed:
Merge branch 'branch-A-BetterGUI'

ryamgoh committed:
Add docs for Ui

ryamgoh committed:
Add config for jekyll theme

ryamgoh committed:
Add changes to build.gradle for win, mac, linux os

ryamgoh committed:
Add Test Cases and Covered Abstraction

ryamgoh committed:
Make changes to header comments

This should cover git conventions
